workflow-ui/core-common/src/main/java/com/squareup/workflow1/ui/Container.kt

```diff
@@ -0,0 +1,46 @@
+package com.squareup.workflow1.ui
+
+import com.squareup.workflow1.ui.Compatible.Companion.keyFor
+
+/**
+ * A model type comprised of a set of other models.
+ *
+ * Why two parameter types? The separate [BaseT] type allows implementations
+ * and sub-interfaces to constrain the types that [map] is allowed to
+ * transform [ContentT] to. E.g., it allows `FooWrapper<S: Screen>` to declare
+ * that [map] is only able to transform `S` to other types of `Screen`.
+ *
+ * @param BaseT the invariant base type of the contents of such a container,
+ * usually [Screen] or [Overlay][com.squareup.workflow1.ui.container.Overlay].
+ * It is common for the [Container] itself to implement [BaseT], but that is
+ * not a requirement. E.g., an `Overlay` might be container of `Screen`.
+ *
+ * @param ContentT the specific subtype of [BaseT] collected by this [Container].
+ */
+@WorkflowUiExperimentalApi
+public interface Container<BaseT, out ContentT : BaseT> {
+  public fun asSequence(): Sequence<ContentT>
+
+  /**
+   * Returns a [Container] with the [transform]ed contents of the receiver.
+   * It is expected that an implementation will take advantage of covariance
+   * to declare its own type as the return type, rather than plain old [Container].
+   * This requirement is not enforced because recursive generics are a fussy nuisance.
+   */
+  public fun <U : BaseT> map(transform: (ContentT) -> U): Container<BaseT, U>
```

Comment:
I _think_ I'm _beginning_ to understand what you are trying to abstract with this `Container` and `map`. It could really use some examples in this kdoc though as its hard to picture.

Reply:
```kotlin
   *
   * For example, suppose we want to create `LoggingScreen`, one that wraps any
   * other screen to add some logging calls. Its implementation of this method
   * would be expected to have a return type of `LoggingScreen` rather than `Container`:
   *
   *    override fun <D : Screen> map(transform: (C) -> D): LoggingScreen<D> =
   *      LoggingScreen(transform(content))
```